File: day_05/day5.jl
#=
day5:
- Julia version: 
- Author: mike.downey
- Date: 2022-12-05
=#
 
data = readlines(open("day_05/data.txt"))

# Split the input into stacks and move instructions
function parse_input(data)
    stacks = []
    moves = []
    part = 0
    for line in data
        if line == ""
            part += 1
            continue
        elseif part == 0
            push!(stacks, line)
        else
            push!(moves, line)
        end
    end
    return stacks,moves
end

remove_char(stacks, howmany) = [s[1 + howmany:end] for s in stacks]
get_leftmost_stack(stacks) = [s[1] for s in stacks if s[1] != ' ']


function build_stacks(stacks)
    cratestacks = []
    while length(stacks[length(stacks)]) > 0
        stacks = remove_char(stacks,1)
        push!(cratestacks, get_leftmost_stack(stacks))
        stacks = remove_char(stacks,3)
    end
    return cratestacks
end

function move_one(cratestacks, wherefrom, whereto)
    crate = popfirst!(cratestacks[wherefrom])
    pushfirst!(cratestacks[whereto], crate)
end

get_moves(instruction) = match(r"(\d+).*(\d+).*(\d+)", instruction).captures

function move_many(cratestacks, wherefrom, whereto, howmany)
    newstack = cratestacks[wherefrom][1:howmany]
    cratestacks[wherefrom] = cratestacks[wherefrom][1 + howmany:end]
    newstack = vcat(newstack, cratestacks[whereto])
    cratestacks[whereto] = newstack
end

# Part 1
println("Part 1")
stacks,moves = parse_input(data)
cratestacks = build_stacks(stacks)
instructions = [[parse(Int, i) for i in get_moves(m)] for m in moves]
for move in instructions
    for i in range(1, move[1])
        move_one(cratestacks, move[2], move[3])
    end
end
println([popfirst!(c) for c in cratestacks])


# Part 2
println("Part 2")
stacks,moves = parse_input(data)
cratestacks = build_stacks(stacks)
for move in instructions
    move_many(cratestacks, move[2], move[3],move[1])
end
println([popfirst!(c) for c in cratestacks])

File: day_05/day5.py
from aoc import aoc
import re


def remove_char(stacks, chars_to_remove=1):
    """traverse down each stack and remove+discard the leftmost item"""
    return [s[chars_to_remove:] for s in stacks]


def get_leftmost_stack(stacks):
    return [s[0] for s in stacks if s[0] != " "]


def build_stacks(stacks):
    cratestacks = []
    while len(stacks[-1]) > 0:
        stacks = remove_char(stacks)
        cratestacks.append(get_leftmost_stack(stacks))
        stacks = remove_char(stacks, 3)
    return cratestacks


def move_one(cratestacks, wherefrom, whereto):
    crate = cratestacks[wherefrom-1].pop(0)
    newstack = [crate]
    newstack.extend(cratestacks[whereto-1])
    cratestacks[whereto-1] = newstack


def part1(stacks, moves):
    cratestacks = build_stacks(stacks)
    instructions = [[int(i) for i in re.findall("\\d+", m)] for m in moves]

    for no, wherefrom, whereto in instructions:
        for i in range(no):
            move_one(cratestacks, wherefrom, whereto)

    return "".join([c[0] for c in cratestacks])


def move_many(cratestacks, wherefrom, whereto, howmany):
    newstack = cratestacks[wherefrom-1][:howmany]
    cratestacks[wherefrom-1] = cratestacks[wherefrom-1][howmany:]
    newstack.extend(cratestacks[whereto-1])
    cratestacks[whereto-1] = newstack


def part2(stacks, moves):
    cratestacks = build_stacks(stacks)
    instructions = [[int(i) for i in re.findall("\\d+", m)] for m in moves]
    for no, wherefrom, whereto in instructions:
        move_many(cratestacks, wherefrom, whereto, no)
    return "".join([c[0] for c in cratestacks])


def run(file):
    print("Part 1")
    data = aoc.loadlines(file)
    stacks, moves = aoc.read_to_blank_line(data)
    print(part1(stacks, moves))
    print("Part 2")
    data = aoc.loadlines(file)
    stacks, moves = aoc.read_to_blank_line(data)
    print(part2(stacks, moves))


if __name__ == "__main__":
    print("Test Data")
    run("test.txt")
    print("Actual Data")
    run("data.txt")
